In [1]:

###### Set Up #####
# verify our folder with the data and module assets is installed
# if it is installed make sure it is the latest
!test -e ds-assets && cd ds-assets && git pull && cd ..
# if it is not installed clone it
!test ! -e ds-assets && git clone https://github.com/IndraniMandal/ds-assets.git
# point to the folder with the assets
home = "ds-assets/assets/"
import sys
sys.path.append(home)

Cloning into 'ds-assets'...
remote: Enumerating objects: 205, done.
remote: Counting objects: 100% (58/58), done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 205 (delta 54), reused 50 (delta 50), pack-reused 147 (from 1)
Receiving objects: 100% (205/205), 12.58 MiB | 4.94 MiB/s, done.
Resolving deltas: 100% (80/80), done.


In [2]:
!ls ds-assets/assets


 2fold-xval.png		      mammals-missing.csv
 5fold-xval.png		     'messy_covid19_southamerica - covid19_southamerica.csv'
 abalone.csv		      mlp.py
 bootstrap.py		      mlp_regression2.py
 caesarian.csv		      mlp_regression.py
 candy-data.csv		      model-performance-curves.png
 cars.csv		      newsgroups.csv
 classification1.jpg	      newsgroups-noheaders.csv
 classification2.jpg	     'Pancreatic Cancer 2020.csv'
 classification3.jpg	      PandasPythonForDataScience.jpg
 colab-badge.afdesign	      PandasPythonForDataScience.pdf
 colab-icon.afdesign	      pdf-badge.png
 colab-icon.png		      perceptron-eq.jpg
 confint.py		      perceptron.jpg
 confusion1.png		      perceptron.r
 confusion2.png		      perceptron-search.png
 crohnd.csv		      perceptron-train.jpg
 cross-validated-curve.png    pipeline.png
 data-science.jpg	      regression1.jpg
 diamonds.csv		      rs.png
 divorce.csv		      shuttle.csv
 divorce-readme.txt	      shuttle.pdf
 elbow.py		      sobar-72.csv
 gapminder_all.c

In [3]:
# set up
import pandas as pd
import numpy as np
np.set_printoptions(formatter={'float_kind':"{:3.2f}".format})
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [4]:
import pandas
url = url = 'https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/crohnd.csv'
df = pd.read_csv(url)
df.head()

ID  nrAdvE    BMI  height  country  sex  age  weight    treat
0  19908       4  25.22     163        0    0   47      67  placebo
1  19909       4  23.80     164        0    0   53      64       d1
2  19910       1  23.05     164        0    0   68      62  placebo
3  20908       1  25.71     165        0    0   48      70       d2
4  20909       2  25.95     170        0    0   67      75  placebo

In [ ]:
df = df.drop(['ID'], axis = 1)

In [ ]:
x = df.drop(['treat'], axis = 1)
y = df['treat']

In [ ]:
# SVM model
model = SVC(kernel='rbf', C=0.001, max_iter=10000)

# do the 5-fold cross validation
scores = cross_val_score(model, x, y, cv=5)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.38 0.38 0.30 0.30 0.30]
Accuracy: 0.33


#Code for the homework

split the data into test and training, check for overfitting, train the data, find the accuracy score

In [5]:
import pandas as pd
import numpy as np
np.set_printoptions(formatter={'float_kind':"{:3.2f}".format})
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from confint import classification_confint

df = pd.read_csv(url)
df = df.drop(['ID'], axis = 1)
x = df.drop(['treat'], axis = 1)
y = df['treat']

actual_y = df['treat']


In [ ]:
# SVM model
model1 = SVC(kernel = 'linear', max_iter=10000)
model2 = SVC(kernel = 'poly', coef0 = 1.0, max_iter=10000)
model3 = SVC(kernel = 'rbf', max_iter=10000)
model4 = SVC(kernel = 'sigmoid', max_iter=10000)

In [ ]:
C = [10**n for n in range(-5,6)]

In [ ]:
model = SVC(max_iter=10000)

In [ ]:
model1.fit(x, actual_y)
model2.fit(x, actual_y)
model3.fit(x, actual_y)
model4.fit(x, actual_y)

/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


SVC(kernel='sigmoid', max_iter=10000)

In [ ]:
linearPrediction = model1.predict(x)
polyPrediction = model2.predict(x)
rbfPrediction = model3.predict(x)
sigmoidPrediction = model4.predict(x)

In [ ]:
acc = accuracy_score(actual_y, linearPrediction)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.44 (0.35,0.53)


In [ ]:
acc = accuracy_score(actual_y, polyPrediction)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.50 (0.41,0.59)


In [ ]:
acc = accuracy_score(actual_y, rbfPrediction)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.39 (0.30,0.48)


In [ ]:
acc = accuracy_score(actual_y, sigmoidPrediction)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.37 (0.28,0.45)


In [ ]:
# grid search
param_grid = [
  {'C': C, 'kernel': ['linear']},
  {'C': C, 'degree': [2, 3, 4, 5, 6], 'coef0': [1.0], 'kernel': ['poly']},
  {'C': C, 'gamma': ['scale', 'auto'], 'kernel': ['rbf']},
  {'C': C, 'coef0': [0.5, 1.0, 1.5], 'kernel': ['sigmoid']},
 ]
grid = GridSearchCV(model, param_grid, cv=5)
grid.fit(x, actual_y)
print("Grid Search: best parameters: {}".format(grid.best_params_))

/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Con

Grid Search: best parameters: {'C': 10000, 'coef0': 1.0, 'degree': 6, 'kernel': 'poly'}


/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


The best parameters are {'C': 10000, 'coef0': 1.0, 'degree': 6, 'kernel': 'poly'}

In [ ]:
# evaluate the best model
best_model = grid.best_estimator_
predict_y = best_model.predict(x)
acc = accuracy_score(actual_y, predict_y)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.41 (0.32,0.50)


In [ ]:
model5 = SVC(kernel = 'sigmoid', coef0 = 0.5, gamma = 'auto', max_iter=10000)
model5.fit(x, actual_y)

predict_y = model5.predict(x)
acc = accuracy_score(actual_y, predict_y)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.33 (0.25,0.42)


In [ ]:
model6 = SVC(kernel = 'sigmoid', coef0 = 0.5, gamma = 'scale', max_iter=10000)
model6.fit(x, actual_y)

predict_y = model6.predict(x)
acc = accuracy_score(actual_y, predict_y)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.37 (0.28,0.45)


In [ ]:
model7 = SVC(kernel = 'poly', coef0 = 0.5, gamma = 'auto', max_iter=10000)
model7.fit(x, actual_y)

predict_y = model7.predict(x)
acc = accuracy_score(actual_y, predict_y)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.32 (0.24,0.41)


/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [ ]:
model8 = SVC(kernel = 'poly', coef0 = 0.5, gamma = 'scale', max_iter=10000)
model8.fit(x, actual_y)

predict_y = model8.predict(x)
acc = accuracy_score(actual_y, predict_y)
lb,ub = classification_confint(acc,x.shape[0])
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Accuracy: 0.44 (0.35,0.53)


In [ ]:
# grid search
param_grid = [
  {'C': C, 'degree': [2, 3, 4, 5, 6], 'coef0': [1.0], 'kernel': ['poly']},
  {'C': C, 'coef0': [0.5, 1.0, 1.5], 'kernel': ['sigmoid']},
 ]

grid = GridSearchCV(model, param_grid, cv=5)
grid.fit(x, actual_y)
print("Grid Search: best parameters: {}".format(grid.best_params_))

/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Con

Grid Search: best parameters: {'C': 10000, 'coef0': 1.0, 'degree': 6, 'kernel': 'poly'}


/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [ ]:
# build the confusion matrix
labels = df['treat'].unique()
cm = confusion_matrix(actual_y, predict_y, labels=labels)
cm_df = pd.DataFrame(cm, index=labels, columns=labels)
print("Confusion Matrix:\n{}".format(cm_df))

Confusion Matrix:
         placebo  d1  d2
placebo       39   0   0
d1            39   0   0
d2            39   0   0


#Answers to the HW

Question: Does tuning the 'gamma' parameter for poly and sigmoid kernels give you better results?

Answer: Poly before 0.32(0.24, 0.41) now 0.44 (0.35, 0.53)

Sigmoid before 0.33(0.25,0.432) now 0.37(0.28, 0.48)

When I changed the gamma parameter from scale to auto for both the poly and sigmoid kernels, the accurcacy and confidence intervals increased giving me better results

Question: Compare your best SVM models to each other. Report if the difference between the models is statistically significant (hint: confidence intervals)

Answer: My best two SVM models are the Linear: 0.44(0.35,0.53) and Poly: 0.50(0.41, 0.59). The difference between these models aren't statistically significant because the confidence interval of 0.53 in Linear overlaps between the confidence intervals of Poly which are 0.41 and 0.59.